### PR TITLE
7.0 ddufresne employee benefits

### PR DESCRIPTION
**Pull request by ghost**
This module implements the employee benefits model for Canada.

### CODE REVIEW

File: l10n_ca_hr_employee_benefits/hr_benefit_category.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp.osv import fields, orm
+
+
+class hr_benefit_category(orm.Model):
+    _name = 'hr.benefit.category'
+    _description = 'Categories of employee benefits'
```
Comment (bwrsandman) [inline l10n_ca_hr_employee_benefits/hr_benefit_category.py]: Please use the singular form.
`_description` is used in many places to describe an object, but most notably in the chattr (Categories of employee benefits created by @dufresnedavid)

File: l10n_ca_hr_employee_benefits/hr_contract_benefit.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+#
+##############################################################################
+
+from openerp.osv import fields, orm
+import time
+
+
+class hr_contract_benefit(orm.Model):
+    _name = 'hr.contract.benefit'
+    _description = 'The benefits in an employee contract'
```
Comment (bwrsandman) [inline l10n_ca_hr_employee_benefits/hr_contract_benefit.py]: Please use singular

File: l10n_ca_hr_employee_benefits/hr_contract.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                    If False, sum over the employee contribution
+
+        annual:     If true, return an annual amount
+                    If false, return a periodic amount
+
+        pays_per_year: The number of pays per year for the employee
+        """
+        # convert string dates to date objects
+        payslip_from = strptime(date_from, "%Y-%m-%d").date()
+        payslip_to = strptime(date_to, "%Y-%m-%d").date()
```
Comment (bwrsandman) [inline l10n_ca_hr_employee_benefits/hr_contract.py]: Wouldn't it be better to use `DEFAULT_SERVER_DATE_FORMAT` here?
`from openerp.tools import DEFAULT_SERVER_DATE_FORMAT`

File: l10n_ca_hr_employee_benefits/hr_contract.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                # Case where the kind of amount requested by the salary rule
+                # is different the amount computed in the benefit record
+                if annual and b.periodicity == 'each_pay':
+                    amount = pays_per_year * amount
+                elif not annual and b.periodicity == 'annual':
+                    amount = amount / pays_per_year
+
+                # Case where the deduction begins after the payslip period
+                # begins.
+                date_start = strptime(b.date_start, "%Y-%m-%d").date()
```
Comment (bwrsandman) [inline l10n_ca_hr_employee_benefits/hr_contract.py]: Wouldn't it be better to use `DEFAULT_SERVER_DATE_FORMAT` here?
`from openerp.tools import DEFAULT_SERVER_DATE_FORMAT`

File: l10n_ca_hr_employee_benefits/hr_contract.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                start_offset = max(
+                    (date_start - payslip_from).days,
+                    0
+                )
+
+                # Case where the deduction ends before the payslip period
+                # ends.
+                date_end = b.date_end and \
+                    strptime(b.date_end, "%Y-%m-%d").date() or False
```
Comment (bwrsandman) [inline l10n_ca_hr_employee_benefits/hr_contract.py]: Wouldn't it be better to use `DEFAULT_SERVER_DATE_FORMAT` here?
`from openerp.tools import DEFAULT_SERVER_DATE_FORMAT`

File: l10n_ca_hr_employee_benefits/hr_contract_benefit.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        ),
+        'code': fields.related(
+            'category_id',
+            'code',
+            type='char',
+            string='Code'
+        ),
+    }
+    _defaults = {
+        'date_start': lambda *a: time.strftime('%Y-%m-%d'),
```
Comment (bwrsandman) [inline l10n_ca_hr_employee_benefits/hr_contract_benefit.py]: Wouldn't it be better to use `DEFAULT_SERVER_DATE_FORMAT` here?
`from openerp.tools import DEFAULT_SERVER_DATE_FORMAT`

File: l10n_ca_hr_employee_benefits/hr_contract_benefit.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        employer_amount=False,
+        name=False,
+        category_id=False
+    ):
+        res = {}
+        if category_id:
+            category = self.pool.get('hr.benefit.category')
+            category = category.browse(cr, uid, category_id)
+            res['value'] = {
+                'employee_amount': employee_amount and employee_amount or
```
Comment (bwrsandman) [inline l10n_ca_hr_employee_benefits/hr_contract_benefit.py]: `employee_amount and employee_amount` is redundant and can be simplified to `employee_amount`

File: l10n_ca_hr_employee_benefits/hr_contract_benefit.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+                category.default_employee_amount,
+                'employer_amount': employer_amount and employer_amount or
```
Comment (bwrsandman) [inline l10n_ca_hr_employee_benefits/hr_contract_benefit.py]: `employee_amount and employee_amount` is redundant and can be simplified to `employee_amount`

File: l10n_ca_hr_employee_benefits/hr_contract_benefit.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+                category.default_employer_amount,
+                'periodicity': category.default_periodicity,
+                'name': name and name or category.name,
```
Comment (bwrsandman) [inline l10n_ca_hr_employee_benefits/hr_contract_benefit.py]: `name and name` is redundant and can be simplified to `name`

File: l10n_ca_hr_employee_benefits/hr_contract.py (hunk 3 above)
Comment (ghost) [inline l10n_ca_hr_employee_benefits/hr_contract.py]: Yes, its much better.

**Comment by bwrsandman** [comment]
Please add a unittests for:
- hr_contact.sum_benefits
- hr_contact_benefit.onchange_category_id

**Comment by bwrsandman** [comment]
Outstanding! :+1:

**Comment by max3903** [comment]
:+1:

**Pull request merged**